# Analysis of Students Performance
=====================================================================================================

This analysis serves as practice to me to learn more in data analytics. This project aims to investigate the key factors that can influence student's performance in exams.

To ensure the flow of the project, this project will follows the 6 steps of data analysis as posted by [DataCamp](https://www.datacamp.com/blog/what-is-data-analysis-expert-guide).

- Step 1: `Defining objectives and questions`
- Step 2: `Data collection`
- Step 3: `Data cleaning`
- Step 4: `Data analysis`
- Step 5: `Data interpretation and visualization`
- Step 6: `Data storytelling`

Step 2 - 4 will be done in `Jupyter` and step 5 - 6 will be done in `PowerBI`

# 1) Defining Objectives and Questions

## Objectives

=========================================================================================================================


**1. Understand Data Distribution:**
   - Analyze the distribution of each numerical variable (e.g., `Hours_Studied`, `Exam_Score`) to identify patterns and trends.

**2. Explore Relationships:**
   - Investigate relationships between different factors (e.g., `Study_Hours` vs. `Exam_Scores`, `Attendance` vs. `Parental_Involvement`) to uncover insights about student performance.

**3. Identify Trends:**
   - Examine trends across categorical variables (e.g., `Teacher_Quality`, `Parental_Education_Level`) and how they correlate with student performance.
  
**4. Analyze Missing Data:**
   - Assess the impact of missing data and its distribution to understand how it may affect analysis.
   
**5. Compare Groups:**
   - Evaluate performance differences between various groups (e.g., `Gender`, `School_Type`, `Parental_Involvement_Level`).

## Questions

**Data Overview:**
- What are the key features in the dataset, and what do they represent?

**Performance Metrics:**
- How do different factors (e.g., hours studied, attendance) influence student exam scores?

**Trends and Patterns:**
- Are there observable trends in student performance based on demographic factors (e.g., gender, parental education)?

**Correlation:**
- Which variables are most strongly correlated with exam scores?

**Impact of Missing Data:**
- What impact does missing data have on the overall analysis, and how should it be addressed?

**Comparison of Groups:**
- How does student performance vary across different groups (e.g., public vs. private schools, different levels of parental involvement)?

**Outliers:**
- Are there any outliers in the dataset, and what might they indicate about the data quality or student behavior?

## Key Metrics

=========================================================================================================================

>`Descriptive Statistics`: `Mean`, `median`, `mode`, `standard deviation`, and `range` for numerical variables (e.g., `Hours_Studied`, `Exam_Score`).

>`Distribution Analysis`: Frequency distributions for categorical variables (e.g., `Parental_Involvement`, `School_Type`).

>`Correlation Coefficients`: Pearson or Spearman correlation coefficients between numerical variables.

>`Group Comparisons`: Performance metrics (e.g., average `Exam_Scores`) segmented by categorical variables (e.g., `Gender`, `Parental_Education_Level`).

>`Missing Data Analysis`: Count and percentage of missing values for each column.

>`Hours_Studied`: Average hours spent studying per week for `high` vs. `low scorers`.

>`Attendance`: Average attendance rates of `high-performing` vs. `low-performing students`.

>`Parental Involvement`: Distribution of students with `High`, `Medium`, and `Low` parental involvement and their corresponding performance.

>`Family Income`: How students from `Low`, `Medium`, and `High` income levels perform in relation to their `Exam_Scores`.

>`School Type`: Performance differences between students attending `Public` vs. `Private` schools.

>`Extracurricular Participation`: Comparison of performance between students who participate in extracurricular activities and those who do not.

>`Teacher Quality`: Correlation between `Teacher Quality` and `Exam Scores`.

## Expected Output

=========================================================================================================================

**Visualizations**
- Various plots (histograms, box plots, scatter plots, etc.) to visually represent the data distributions and relationships.

**Descriptive Summary**

- Generate statistical summaries (e.g., mean, median, mode) for key attributes like `Exam_Score`, `Hours_Studied`, `Attendance`, etc.
- Visualize distributions of key factors (e.g., bar charts for `Family Income` vs. `Exam_Score`).

**Correlation Analysis:**

- Identify correlations between variables such as `Hours_Studied`, `Parental Involvement`, and `Exam_Score` to understand which factors are more closely tied to performance.

**Group Comparison:**

- Break down Exam_Scores by different demographic groups (e.g., `School_Type`, `Parental_Involvement`, `Family_Income`).
- Identify performance gaps across different groups.

**Insights and Recommendations:**

- Based on the data, provide recommendations for educational interventions (e.g., increasing `Tutoring_Sessions` for students with Low `Parental Involvement` or improving `Access_to_Resources` for students from low-income families).

# 2) Data Collection

The data was taken from a Kaggle repository. It was uploaded by [Practice Data Analysis With Me](https://www.kaggle.com/lainguyn123). If you're interested in doing the analysis yourself, you can download the data [here](https://www.kaggle.com/datasets/lainguyn123/student-performance-factors). In this step, we will look into the dataset's properties.

### 2.1) Importing the data

In [1]:
import pandas as pd

df = pd.read_csv("StudentPerformanceFactors.csv")

### 2.2) Checking data types

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6607 entries, 0 to 6606
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Hours_Studied               6607 non-null   int64 
 1   Attendance                  6607 non-null   int64 
 2   Parental_Involvement        6607 non-null   object
 3   Access_to_Resources         6607 non-null   object
 4   Extracurricular_Activities  6607 non-null   object
 5   Sleep_Hours                 6607 non-null   int64 
 6   Previous_Scores             6607 non-null   int64 
 7   Motivation_Level            6607 non-null   object
 8   Internet_Access             6607 non-null   object
 9   Tutoring_Sessions           6607 non-null   int64 
 10  Family_Income               6607 non-null   object
 11  Teacher_Quality             6529 non-null   object
 12  School_Type                 6607 non-null   object
 13  Peer_Influence              6607 non-null   obje

It seems that there are missing data several columns in the dataset. We will deal with this in step 3 later on.

### 2.3) Viewing top portion of the dataset

In [3]:
df.head()

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0             23          84                  Low                High   
1             19          64                  Low              Medium   
2             24          98               Medium              Medium   
3             29          89                  Low              Medium   
4             19          92               Medium              Medium   

  Extracurricular_Activities  Sleep_Hours  Previous_Scores Motivation_Level  \
0                         No            7               73              Low   
1                         No            8               59              Low   
2                        Yes            7               91           Medium   
3                        Yes            8               98           Medium   
4                        Yes            6               65           Medium   

  Internet_Access  Tutoring_Sessions Family_Income Teacher_Quality  \
0             Yes                  0           Low          Medium   
1             Yes                  2        Medium          Medium   
2             Yes                  2        Medium          Medium   
3             Yes                  1        Medium          Medium   
4             Yes                  3        Medium            High   

  School_Type Peer_Influence  Physical_Activity Learning_Disabilities  \
0      Public       Positive                  3                    No   
1      Public       Negative                  4                    No   
2      Public        Neutral                  4                    No   
3      Public       Negative                  4                    No   
4      Public        Neutral                  4                    No   

  Parental_Education_Level Distance_from_Home  Gender  Exam_Score  
0              High School               Near    Male          67  
1                  College           Moderate  Female          61  
2             Postgraduate               Near    Male          74  
3              High School           Moderate    Male          71  
4                  College               Near  Female          70

Since the naming of the columns are consistent, we won't be changing the column names

### 2.4) Data shape

In [4]:
df_before = df.shape
print(f'Shape of the data: {df_before}')

Shape of the data: (6607, 20)


### 2.5) Statistical description

This step is important as it can show us of there is any outliers in data. It can also help us to decide on how to replace the missing value either using mean, mode, or median value of the data to reduce biases.

In [5]:
#Calculate the mean, min, max, and other descriptive statistical value 
df.describe(include='all')

Hours_Studied   Attendance Parental_Involvement Access_to_Resources  \
count     6607.000000  6607.000000                 6607                6607   
unique            NaN          NaN                    3                   3   
top               NaN          NaN               Medium              Medium   
freq              NaN          NaN                 3362                3319   
mean        19.975329    79.977448                  NaN                 NaN   
std          5.990594    11.547475                  NaN                 NaN   
min          1.000000    60.000000                  NaN                 NaN   
25%         16.000000    70.000000                  NaN                 NaN   
50%         20.000000    80.000000                  NaN                 NaN   
75%         24.000000    90.000000                  NaN                 NaN   
max         44.000000   100.000000                  NaN                 NaN   

       Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
count                        6607   6607.00000      6607.000000   
unique                          2          NaN              NaN   
top                           Yes          NaN              NaN   
freq                         3938          NaN              NaN   
mean                          NaN      7.02906        75.070531   
std                           NaN      1.46812        14.399784   
min                           NaN      4.00000        50.000000   
25%                           NaN      6.00000        63.000000   
50%                           NaN      7.00000        75.000000   
75%                           NaN      8.00000        88.000000   
max                           NaN     10.00000       100.000000   

       Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
count              6607            6607        6607.000000          6607   
unique                3               2                NaN             3   
top              Medium             Yes                NaN           Low   
freq               3351            6108                NaN          2672   
mean                NaN             NaN           1.493719           NaN   
std                 NaN             NaN           1.230570           NaN   
min                 NaN             NaN           0.000000           NaN   
25%                 NaN             NaN           1.000000           NaN   
50%                 NaN             NaN           1.000000           NaN   
75%                 NaN             NaN           2.000000           NaN   
max                 NaN             NaN           8.000000           NaN   

       Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
count             6529        6607           6607        6607.000000   
unique               3           2              3                NaN   
top             Medium      Public       Positive                NaN   
freq              3925        4598           2638                NaN   
mean               NaN         NaN            NaN           2.967610   
std                NaN         NaN            NaN           1.031231   
min                NaN         NaN            NaN           0.000000   
25%                NaN         NaN            NaN           2.000000   
50%                NaN         NaN            NaN           3.000000   
75%                NaN         NaN            NaN           4.000000   
max                NaN         NaN            NaN           6.000000   

       Learning_Disabilities Parental_Education_Level Distance_from_Home  \
count                   6607                     6517               6540   
unique                     2                        3                  3   
top                       No              High School               Near   
freq                    5912                     3223               3884   
mean                     NaN                      NaN                NaN   
std                 

In [6]:
#Calculate the mode of the data
df.mode()

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0             20          67               Medium              Medium   

  Extracurricular_Activities  Sleep_Hours  Previous_Scores Motivation_Level  \
0                        Yes            7               66           Medium   

  Internet_Access  Tutoring_Sessions Family_Income Teacher_Quality  \
0             Yes                  1           Low          Medium   

  School_Type Peer_Influence  Physical_Activity Learning_Disabilities  \
0      Public       Positive                  3                    No   

  Parental_Education_Level Distance_from_Home Gender  Exam_Score  
0              High School               Near   Male          68

In [7]:
#Calculate the median of the numerical data
df.median(numeric_only=True)

Hours_Studied        20.0
Attendance           80.0
Sleep_Hours           7.0
Previous_Scores      75.0
Tutoring_Sessions     1.0
Physical_Activity     3.0
Exam_Score           67.0
dtype: float64

### Summary

**Numerical**

`Hours_Studied`
- mean: 19.975329
- median: 20.0
- mode: 20
- std: 5.990594

`Attendance`
- mean: 79.977448
- median: 80.0
- mode: 67
- std: 11.547475	

`Sleep_Hours`
- mean: 7.02906
- median: 7.0
- mode: 7
- std: 1.46812	

`Previous_Scores`
- mean: 75.070531
- median: 75.0
- mode: 66
- std: 14.399784	

`Tutoring_Sessions`
- mean: 1.493719
- median: 1.0
- mode: 1
- std: 1.230570	

`Physical_Activity`
- mean: 2.967610
- median: 3.0
- mode: 3
- std: 1.031231	

`Exam_Score`
- mean: 67.235659
- median: 67.0
- mode: 68
- std: 3.890456

**String**
			

`Parental_Involvement`
- frequency: 3362
- mode: Medium

`Access_to_Resources`
- frequency: 3319
- mode: Medium 

`Extracurricular_Activities`
- frequency: 3938
- mode: Yes

`Motivation_Level`
- frequency: 3351
- mode: Medium

`Internet_Access`
- frequency: 6108
- mode: Yes

`Family_Income`
- frequency: 2672
- mode: Low

`Teacher_Quality`
- frequency: 3925
- mode: Medium

`School_Type`
- frequency: 4598
- mode: Public

`Peer_Influence`
- frequency: 2638
- mode: Positive

`Learning_Disabilities`
- frequency: 5912
- mode: No

`Parental_Education_Level`
- frequency: 3223
- mode: High School

`Distance_from_Home`
- frequency: 3884	
- mode: Near

`Gender`
- frequency: 3814
- mode: Male

## 3) Data Cleaning

### 3.1) Checking for duplicates entries

In [8]:
print(f'Total duplicates: \n{df.duplicated().sum()}')

Total duplicates: 
0


### 3.2) Checking for missing/null value

In [9]:
df.isnull().sum()
print(f'DataFrame displaying False for Null (NaN) value = \n{df.isnull().sum()}')

DataFrame displaying False for Null (NaN) value = 
Hours_Studied                  0
Attendance                     0
Parental_Involvement           0
Access_to_Resources            0
Extracurricular_Activities     0
Sleep_Hours                    0
Previous_Scores                0
Motivation_Level               0
Internet_Access                0
Tutoring_Sessions              0
Family_Income                  0
Teacher_Quality               78
School_Type                    0
Peer_Influence                 0
Physical_Activity              0
Learning_Disabilities          0
Parental_Education_Level      90
Distance_from_Home            67
Gender                         0
Exam_Score                     0
dtype: int64


As shown here and previously in step 2.2, there are 3 columns with missing data. Considering the percentage of missing data is less than 10% of total rows, the rows with missing data will dropped.

### 3.3) Dropping rows with missing value

In [10]:
df = df.dropna()

In [11]:
#Verifying that the missing value have been drop
print(df.isnull().sum())

Hours_Studied                 0
Attendance                    0
Parental_Involvement          0
Access_to_Resources           0
Extracurricular_Activities    0
Sleep_Hours                   0
Previous_Scores               0
Motivation_Level              0
Internet_Access               0
Tutoring_Sessions             0
Family_Income                 0
Teacher_Quality               0
School_Type                   0
Peer_Influence                0
Physical_Activity             0
Learning_Disabilities         0
Parental_Education_Level      0
Distance_from_Home            0
Gender                        0
Exam_Score                    0
dtype: int64


In [12]:
# Number of deleted rows rows after cleaning
df_after = df.shape
print(f'Number of deleted rows: {df_before[0] - df_after[0]}')

Number of deleted rows: 229


In [13]:
#Number of rows after dropping rows with null values
print(f'Number of rows: {df_after[0]}')

Number of rows: 6378


### 3.4) Removing the outliers 

We will be using IQR for detecting and removing numerical outliers from the dataset

In [14]:
# Define numerical columns
numerical_columns = ['Hours_Studied', 'Attendance', 'Sleep_Hours', 'Previous_Scores', 'Tutoring_Sessions', 'Physical_Activity', 'Exam_Score']

# Function to detect outliers using IQR and remove them
def remove_outliers_iqr(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    # Removing the rows that contain outliers
    df_cleaned = df[(df[column] >= lower_bound) & (df[column] <= upper_bound)]
    return df_cleaned

# Apply the function to all numerical columns
for column in numerical_columns:
    df = remove_outliers_iqr(df, column)

In [15]:
# Print the cleaned DataFrame shape to check how many rows remain
print("DataFrame after removing numerical outliers:", df.shape)

DataFrame after removing numerical outliers: (5836, 20)


### 3.5) Checking the statistical data of the cleaned data

In [18]:
df.describe(include='all')

Hours_Studied   Attendance Parental_Involvement Access_to_Resources  \
count     5836.000000  5836.000000                 5836                5836   
unique            NaN          NaN                    3                   3   
top               NaN          NaN               Medium              Medium   
freq              NaN          NaN                 2992                2936   
mean        20.008053    80.021933                  NaN                 NaN   
std          5.782292    11.499526                  NaN                 NaN   
min          4.000000    60.000000                  NaN                 NaN   
25%         16.000000    70.000000                  NaN                 NaN   
50%         20.000000    80.000000                  NaN                 NaN   
75%         24.000000    90.000000                  NaN                 NaN   
max         36.000000   100.000000                  NaN                 NaN   

       Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
count                        5836  5836.000000      5836.000000   
unique                          2          NaN              NaN   
top                           Yes          NaN              NaN   
freq                         3480          NaN              NaN   
mean                          NaN     7.041295        75.107779   
std                           NaN     1.469662        14.348109   
min                           NaN     4.000000        50.000000   
25%                           NaN     6.000000        63.000000   
50%                           NaN     7.000000        75.000000   
75%                           NaN     8.000000        87.000000   
max                           NaN    10.000000       100.000000   

       Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
count              5836            5836        5836.000000          5836   
unique                3               2                NaN             3   
top              Medium             Yes                NaN        Medium   
freq               2967            5387                NaN          2364   
mean                NaN             NaN           1.291809           NaN   
std                 NaN             NaN           0.982455           NaN   
min                 NaN             NaN           0.000000           NaN   
25%                 NaN             NaN           1.000000           NaN   
50%                 NaN             NaN           1.000000           NaN   
75%                 NaN             NaN           2.000000           NaN   
max                 NaN             NaN           3.000000           NaN   

       Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
count             5836        5836           5836        5836.000000   
unique               3           2              3                NaN   
top             Medium      Public       Positive                NaN   
freq              3499        4061           2332                NaN   
mean               NaN         NaN            NaN           2.967443   
std                NaN         NaN            NaN           1.026956   
min                NaN         NaN            NaN           0.000000   
25%                NaN         NaN            NaN           2.000000   
50%                NaN         NaN            NaN           3.000000   
75%                NaN         NaN            NaN           4.000000   
max                NaN         NaN            NaN           6.000000   

       Learning_Disabilities Parental_Education_Level Distance_from_Home  \
count                   5836                     5836               5836   
unique                     2                        3                  3   
top                       No              High School               Near   
freq                    5241                     2892               3479   
mean                     NaN                      NaN                NaN   
std                 

In [19]:
df.mode()

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0             20          67               Medium              Medium   

  Extracurricular_Activities  Sleep_Hours  Previous_Scores Motivation_Level  \
0                        Yes            7               66           Medium   

  Internet_Access  Tutoring_Sessions Family_Income Teacher_Quality  \
0             Yes                  1        Medium          Medium   

  School_Type Peer_Influence  Physical_Activity Learning_Disabilities  \
0      Public       Positive                  3                    No   

  Parental_Education_Level Distance_from_Home Gender  Exam_Score  
0              High School               Near   Male          66

In [20]:
df.median(numeric_only=True)

Hours_Studied        20.0
Attendance           80.0
Sleep_Hours           7.0
Previous_Scores      75.0
Tutoring_Sessions     1.0
Physical_Activity     3.0
Exam_Score           67.0
dtype: float64

### 3.6) Exporting the cleaned data

In [21]:
df.to_csv('studentPerformance_clean.csv', index=False)